### PR TITLE
Revoke endpoint refactoring and fix some issues

### PR DESCRIPTION
**Pull request by IsuraD**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

- Refactor revoke endpoint
- Fixed issues related to repeated parameters
- Fixed issues related to error response when required parameters missing

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

- Now
### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

- No

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.endpoint.exception;
+
+public class RevokeEndpointAccessDeniedException extends InvalidRequestParentException {
+
+    String callback;
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java]: Private

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.oauth.endpoint.exception;
+
+public class RevokeEndpointAccessDeniedException extends InvalidRequestParentException {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java]: Better to have a descriptive comment even for Exceptions

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/expmapper/InvalidRequestExceptionMapper.java [hunk 3]
```diff
@@ -155,4 +170,26 @@ private Response buildErrorResponse(int status, InvalidRequestParentException ex
...
+
+        String callback = ((RevokeEndpointAccessDeniedException) exception).getCallback();
+        if (isBlank(callback)) {
+            OAuthResponse response = OAuthASResponse.errorResponse(HttpServletResponse.SC_UNAUTHORIZED)
+                    .setError(OAuth2ErrorCodes.INVALID_CLIENT)
+                    .setErrorDescription(exception.getMessage()).buildJSONMessage();
+
+            return Response.status(response.getResponseStatus())
+                    .header(OAuthConstants.HTTP_RESP_HEADER_AUTHENTICATE, EndpointUtil.getRealmInfo())
+                    .header(HttpHeaders.CONTENT_TYPE, "text/html")
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/expmapper/InvalidRequestExceptionMapper.java]: Do we have any constant for "text/html" or  "application/javascript"?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/revoke/OAuthRevocationEndpoint.java [hunk 4]
```diff
@@ -28,271 +27,330 @@
...
     }
 
-    private OAuthRevocationResponseDTO revokeTokens(OAuthRevocationRequestDTO oauthRequest)
-            throws OAuthClientException {
+    private OAuthRevocationResponseDTO revokeTokens(OAuthRevocationRequestDTO oauthRequest) {
+        OAuthRevocationRequestDTO revokeReqDTO = getoAuthRevocationRequestDTO(oauthRequest);
+        return getOAuth2Service().revokeTokenByOAuthClient(revokeReqDTO);
+    }
 
+    private OAuthRevocationRequestDTO getoAuthRevocationRequestDTO(OAuthRevocationRequestDTO oauthRequest) {
```
Comment (nuwandiw) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/revoke/OAuthRevocationEndpoint.java]: Looks like this method is redundant. It takes **OAuthRevocationRequestDTO** object as parameter, build another **OAuthRevocationRequestDTO** object with details of the sent parameter. Can we remove this method and directly call getOAuth2Service().revokeTokenByOAuthClient(oauthRequest) in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/667/files#diff-c9fcb5ce9b0a94b5aa9ac6891ba5cba9R337

Comment (IsuraD) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/revoke/OAuthRevocationEndpoint.java]: Yes. Fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java (hunk 1 above)
Comment (IsuraD) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java]: fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java (hunk 2 above)
Comment (IsuraD) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/exception/RevokeEndpointAccessDeniedException.java]: done

**Pull request merged**